Documentation.app/Contents/MacOS/FP concepts.playground/Pages/Rank-N polymorphism.xcplaygroundpage/Contents.swift

```diff
@@ -0,0 +1,155 @@
+// nef:begin:header
+/*
+ layout: docs
+ title: Rank-N polymorphism
+ */
+// nef:end
+// nef:begin:hidden
+import Bow
+// nef:end
+/*:
+ # Rank-N polymorphism
+
+ {:.intermediate}
+ intermediate
+
+ Swift does not support Rank-N polymorphism yet. However, Rank-N polymorphism unlocks some interesting patterns and techniques. For this reason, Bow provides some helper classes that allow you to emulate Rank-N polymorphism. This document introduces Rank-N polymorphism and Bow's alternative approach starting from the basics of closures.
+
+ ## What is Rank-N polymorphism?
+
+ Swift allows us to write generic functions: functions that abstract over what specific types they operate on.
+*/
+func singletonArray<T>(_ v: T) -> [T] {
+    [v]
+}
+/*:
+ Here `singletonArray` is a function that can operate on any type we want. Specifically, the caller of the function decides what type to pass to the function.
+
+ We say that such a function is *polymorphic* over the *type parameter* `T`, or alternatively that the function is *generic* over the type parameter `T`. Notice that we don't refer to `T` as a type. `T` is more like a placeholder for a type that will be determined when the function is called.
+
+ Imagine we want to write a function `callTwice` that is generic over two type parameter `A` and `B` and accepts our `singletonArray` above as a parameter. Our function `callTwice` will execute the polymorphic function we pass to it with a value of type `A` and a second time with a value of type `B`. We need to express that `callTwice` takes a parameter that is a function polymorphic on a type parameter `T`. The function could look something like this:
+ ```
+ func callTwice<A, B>(_ f: <T>(T) -> [T], _ a: A, _ b: B) -> ([A], [B]) {
+     (f(a), f(b)) // `f` really needs to be polymorphic, because we call it twice with inputs of different types.
+ }
+ ```
+ Notice that we included `<T>` in front of the type signature of the parameter of our function `callTwice`. With this, we want to express that the parameter `f` must be a function polymorphic over one type parameter we called `T`. This is a made up syntax, don't try to write this in your swift program because it won't compile.
```

Comment:
```suggestion
 Notice that we included `<T>` in front of the type signature of the parameter of our function `callTwice`. With this, we want to express that the parameter `f` must be a function polymorphic over one type parameter we called `T`. This is a made up syntax, don't try to write this in your Swift program because it won't compile.
```